File: webapp/src/main/scala/org/linnando/mixemulator/webapp/LineAccessFileOutputOpsImpl.scala
package org.linnando.mixemulator.webapp

import io.scalajs.nodejs.buffer.Buffer
import io.scalajs.nodejs.fs.{FileInputOptions, Fs}
import org.linnando.mixemulator.vm.io.file.LineAccessFileOutputOps

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future

class LineAccessFileOutputOpsImpl extends LineAccessFileOutputOps {
  override def appendLine(filename: String, version: Int, chars: Array[Char]): Future[Unit] = {
    val oldFile = s"/$filename/$version"
    val newFile = s"/$filename/${version + 1}"
    for {
      exists <- Fs.existsFuture(oldFile)
      _ <- if (exists) copyFileFuture(oldFile, newFile) else Future {}
      _ <- appendString(newFile, chars.mkString + "\n")
    } yield ()
  }

  // TODO Implement without reading the whole file to memory
  private def copyFileFuture(src: String, dest: String): Future[Unit] =
    for {
      buffer <- Fs.readFileFuture(src).map(_.asInstanceOf[Buffer])
      _ <- Fs.writeFileFuture(dest, buffer)
    } yield ()

  private def appendString(filename: String, str: String): Future[Unit] = {
    Fs.openFuture(filename, "a").flatMap(dest => {
      Fs.writeFuture(dest, str).map(_ => ()) andThen {
        case _ => Fs.closeFuture(dest)
      }
    })
  }

  override def appendNewPage(filename: String, version: Int): Future[Unit] = {
    val oldFile = s"/$filename/$version"
    val newFile = s"/$filename/${version + 1}"
    for {
      exists <- Fs.existsFuture(oldFile)
      _ <- if (exists) copyFileFuture(oldFile, newFile) else Future {}
      _ <- appendString(newFile, "\f")
    } yield ()
  }

  override def initialise(filename: String): Future[Unit] = for {
    _ <- ensureDeviceDirectoryExists(filename)
    versions <- getVersions(filename)
    _ <- Fs.writeFileFuture(s"/$filename/0", "")
    _ <- unlinkVersions(filename, versions.filter(_ != "0"))
  } yield ()

  override def getVersions(filename: String): Future[Iterable[String]] =
    Fs.readdirFuture(s"/$filename").map(_.toArray[String])

  private def ensureDeviceDirectoryExists(filename: String): Future[Unit] = for {
    exists <- Fs.existsFuture(s"/$filename")
    isDirectory <- if (exists) Fs.statFuture(s"/$filename").map(_.isDirectory()) else Future(false)
    _ <- if (exists && !isDirectory) Fs.unlinkFuture(s"/$filename") else Future {}
    _ <- if (exists && isDirectory) Future {} else Fs.mkdirFuture(s"/$filename")
  } yield ()

  private def unlinkVersions(filename: String, versions: Iterable[String]): Future[Unit] = {
    val futures = versions.map(version => Fs.unlinkFuture(s"/$filename/$version"))
    Future.fold(futures)(())((_, _) => ())
  }

  override def getData(filename: String, version: Int): Future[String] =
    Fs.readFileFuture(s"/$filename/$version", new FileInputOptions(encoding = "utf8")).map(_.asInstanceOf[String])
}

File: webapp/src/app/line-access-file-output-ops-impl.ts
import { LineAccessFileOutputOps } from '@mixemulator/lib';

export class LineAccessFileOutputOpsImpl extends LineAccessFileOutputOps {
  constructor(private storage: Storage) {
    super();
  }

  override appendLine(filename: string, version: number, chars: string): Promise<void> {
    return this.appendCharsToNewVersion(filename, version, chars + '\n');
  }

  private appendCharsToNewVersion(filename: string, version: number, chars: string): Promise<void> {
    const oldFile = `${filename}/${version}`;
    const newFile = `${filename}/${version + 1}`;
    return new Promise((resolve, reject) => {
      try {
        const srcData = this.storage.getItem(oldFile) ?? '';
        this.storage.setItem(newFile, srcData + chars);
        return resolve();
      } catch (e) {
        return reject(e);
      }
    });
  }

  override appendNewPage(filename: string, version: number): Promise<void> {
    return this.appendCharsToNewVersion(filename, version, 'f');
  }

  override async initialise(filename: string): Promise<void> {
    const versions = await this.getVersions(filename);
    return new Promise((resolve, reject) => {
      try {
        for (const version of versions) {
          this.storage.removeItem(`${filename}/${version}`);
        }
        this.storage.setItem(`${filename}/0`, '');
        return resolve();
      } catch (e) {
        return reject(e);
      }
    });
  }

  override getVersions(filename: string): Promise<string[]> {
    return new Promise((resolve, reject) => {
      try {
        const versions: string[] = [];
        for (let i = 0; i < this.storage.length; i++) {
          const key = this.storage.key(i);
          const match = key?.match(new RegExp(`^${filename}/(\\d+)$`));
          if (match) {
            versions.push(match[1]);
          }
        }
        return resolve(versions);
      } catch (e) {
        return reject(e);
      }
    });
  }

  override getData(filename: string, version: number): Promise<string> {
    return new Promise((resolve, reject) => {
      try {
        const data = this.storage.getItem(`${filename}/${version}`);
        if (data == null) {
          return reject(new Error(`File ${filename}/${version} not found.`));
        }
        return resolve(data);
      } catch (e) {
        return reject(e);
      }
    });
  }
}
